lib/clients/ZookeeperManager.js

```diff
@@ -0,0 +1,323 @@
+     *
+     * If 'autoCreateNamespace' is set in the options and the base path does not exist in ZooKeeper,
+     * this method tries to create the necessary path (namespace) in ZooKeeper.
+     *
+     * @method _connect
+     * @return {undefined}
+     */
+    _connect() {
+        // clean up exists client before reconnect
+        if (this.client) {
+            // close and clean up the existing ZooKeeper client connection.
+            this.close();
+            this.client.removeAllListeners();
+        }
+
+        this.client = zookeeper.createClient(this.connectionString, this.options);
+
+        let nsIndex;
+        let namespace;
+        let rootZkClient;
+        async.series([
+            // Check zookeeper client is connected
+            next => this.client.once('connected', next),
+            // Create namespace if it exists and options.autoCreateNamespace is true
+            next => {
+                // TODO: ARTESCA-10337 The 'autoCreateNamespace' functionality is currently specific to
+                // Artesca and may be removed in future versions once the Zenko Operator can handle base path creation.
+                // Once removed, we can simply rely on the 'connected' state instead of the 'ready' state and
+                // stop listening on 'error' event.
+                if (this.options && this.options.autoCreateNamespace) {
+                    nsIndex = this.connectionString.indexOf('/');
+                    namespace = this.connectionString.slice(nsIndex);
+                    if (nsIndex > -1 && namespace !== '/') {
+                        return process.nextTick(next);
+                    }
+                }
+                return process.nextTick(() => next({ event: 'ready' }));
+            },
+            // Check that the namespace is not already created.
+            next => this.getData('/', err => {
+                if (err && err.name !== 'NO_NODE') {
+                    return next({ event: 'error', err });
+                }
+                // If not created (NO_NODE error), go to next step.
+                if (err) {
+                    return next();
+                }
+                return next({ event: 'ready' });
+            }),
+            // Since the namespace has not been created, connect to the root zookeeper client
+            next => {
+                const hostPort = this.connectionString.slice(0, nsIndex);
+                rootZkClient = zookeeper.createClient(hostPort, this.options);
+                rootZkClient.connect();
+                rootZkClient.once('connected', next);
+            },
+            // Once connected, use the root zookeeper client to create the namespace
+            next => rootZkClient.mkdirp(namespace, err => {
+                if (err && err.name !== 'NODE_EXISTS') {
+                    return next({ event: 'error', err });
+                }
+                return next({ event: 'ready' });
```

Comment:
Do you need to disconnect/close the `rootZkClient` once the path exists?

Reply:
Good catch. Thanks!